myst_nbs/time_series/Forecasting_with_structural_timeseries.myst.md

```diff
@@ -0,0 +1,747 @@
+
+```{code-cell} ipython3
+priors_0 = {
+    "coefs": {"mu": [-4, 0.2], "sigma": 0.1, "size": 2},
+    "sigma": 8,
+    "init": {"mu": 9, "sigma": 0.1, "size": 1},
+}
+
+priors_1 = {
+    "coefs": {"mu": [-2, 0.2], "sigma": 0.1, "size": 2},
+    "sigma": 12,
+    "init": {"mu": 8, "sigma": 0.1, "size": 1},
+}
+
+priors_2 = {
+    "coefs": {"mu": [0, 0.2], "sigma": 0.1, "size": 2},
+    "sigma": 15,
+    "init": {"mu": 8, "sigma": 0.1, "size": 1},
+}
+
+models = {}
+for i, p in enumerate([priors_0, priors_1, priors_2]):
+    models[i] = {}
+    idata = make_latent_AR_model(y_t, p, full_sample=False)
+    models[i]["idata"] = idata
+```
+
+```{code-cell} ipython3
+fig, axs = plt.subplots(1, 3, figsize=(10, 4), sharey=True)
+axs = axs.flatten()
+for i, p in zip(range(3), [priors_0, priors_1, priors_2]):
+    axs[i].plot(
+        az.extract_dataset(models[i]["idata"], group="prior_predictive", num_samples=100)[
+            "likelihood"
+        ],
+        color="blue",
+        alpha=0.1,
+    )
+    axs[i].plot(y_t, "o", color="black", markersize=2)
+    axs[i].set_title(
+        "$y_{t+1}$" + f'= N({p["coefs"]["mu"][0]} + {p["coefs"]["mu"][1]}y$_t$, {p["sigma"]})'
+    )
+plt.suptitle("Prior Predictive Specifications", fontsize=20);
+```
+
+We can see the manner in which the model struggles to capture the trend line. Increasing the variablity of the model will never capture the directional pattern we know to be in the data.
+
+```{code-cell} ipython3
+priors_0 = {
+    "coefs": {"mu": [-4, 0.2], "sigma": [0.5, 0.03], "size": 2},
+    "sigma": 8,
+    "init": {"mu": -4, "sigma": 0.1, "size": 1},
+}
+
+idata_no_trend, preds_no_trend, model = make_latent_AR_model(y_t, priors_0)
+```
+
+```{code-cell} ipython3
+plot_fits(idata_no_trend, preds_no_trend)
+```
+
+Forecasting with this model is somewhat hopeless because, while the model fit adjusts well with observed data, but it completely fails to capture the structural trend in the data. So without some structural constraint when we seek to make predictions with this simple AR model, it reverts to the mean level forecast very quickly.
+
++++
+
+### Specifying a Trend Model
+
+We will define a model to account for the trend in our data and combine this trend in a additive model with the autoregressive components. Again the model is much as before, but now we add additional latent features. These are to be combined in a simple additive combination but we can be more creative here if it would suit our model.
+
+```{code-cell} ipython3
+def make_latent_AR_trend_model(
+    ar_data, priors, prediction_steps=250, full_sample=True, samples=2000
+):
+    with pm.Model() as AR:
+        pass
+
+    t_data = list(range(len(ar_data)))
+    AR.add_coord("obs_id", t_data, mutable=True)
```

Comment:
Same suggestion about coords here. I think the most important would be adding `obs_id` as a dim to the `ar` and `likelihood` nodes.

But it would also be good in this model to add a coord for the length 50 variable. At the moment the resulting graphviz is a bit disjointed because it's not matching up nodes in plates.

Reply:
Done